AK/CircularBuffer.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2022, Lucas Chollet <lucas.chollet@free.fr>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CircularBuffer.h>
+#include <AK/MemMem.h>
+
+namespace AK {
+
+CircularBuffer::CircularBuffer(ByteBuffer buffer)
+    : m_buffer(move(buffer))
+{
+}
+
+ErrorOr<CircularBuffer> CircularBuffer::create_empty(size_t size)
+{
+    auto temporary_buffer = TRY(ByteBuffer::create_uninitialized(size));
+
+    CircularBuffer circular_buffer { move(temporary_buffer) };
+
+    return circular_buffer;
+}
+
+ErrorOr<CircularBuffer> CircularBuffer::create_initialized(ByteBuffer buffer)
+{
+    CircularBuffer circular_buffer { move(buffer) };
+
+    circular_buffer.m_used_space = circular_buffer.m_buffer.size();
+
+    return circular_buffer;
+}
+
+size_t CircularBuffer::empty_space() const
+{
+    return capacity() - m_used_space;
+}
+
+size_t CircularBuffer::used_space() const
+{
+    return m_used_space;
+}
+
+size_t CircularBuffer::capacity() const
+{
+    return m_buffer.size();
+}
+
+bool CircularBuffer::is_wrapping_around() const
+{
+    return capacity() <= m_reading_head + m_used_space;
+}
+
+Optional<size_t> CircularBuffer::offset_off(StringView needle, Optional<size_t> until) const
+{
+    Array<ReadonlyBytes, 2> spans {};
+    spans[0] = next_read_span();
+
+    if (until.has_value() && spans[0].size() > *until) {
+        spans[0] = spans[0].trim(*until);
+    } else {
+        if (is_wrapping_around())
+            spans[1] = m_buffer.span().slice(0, (m_reading_head + m_used_space) % capacity());
+        if (until.has_value() && spans[0].size() + spans[1].size() > *until)
+            spans[1] = spans[1].trim(*until - spans[0].size());
+    }
```

Comment:
Suggestion: Do `size_t real_until = until.value_or(m_used_size)` (or something similar to that, naming suggestions are also welcome), and then work with `real_until` unconditionally.

The code should boil down to the following then:

```suggestion
    size_t real_until = until.value_or(m_used_space);
    Array<ReadonlyBytes, 2> spans {};
    spans[0] = next_read_span();

    if (spans[0].size() > real_until)
        spans[0] = spans[0].trim(real_until);

    if (is_wrapping_around())
        spans[1] = m_buffer.span().slice(0, real_until - spans[0].size());
```